### ISSUE TITLE
Disconnects on long wait time for new messages

### ISSUE THREAD

**Issue by SimplicityGuy** [sbtinstruments/aiomqtt#27]
I have a discord bot that is supposed to send a message to discord after receiving an MQTT message. There is very little activity, so typically there is hours or even days between messages received. I noticed that when there's a long wait, the first time a message is published and is received by the bot, I hit `Disconnected during message iteration`. The result is the message on the subscribed topic is gone. Here is a log of the output when this happens:
```
bot   | 2020-12-19 05:51:55 INFO Connecting to MQTT.
bot   | 2020-12-19 05:51:55 INFO Connection to MQTT open.
...
bot   | 2020-12-19 16:18:19 ERROR Connection to MQTT closed: Disconnected during message iteration
bot   | 2020-12-19 16:18:22 INFO Connecting to MQTT.
bot   | 2020-12-19 16:18:22 INFO Connection to MQTT open.
```
Subsequent messages are published minutes later, and the bot is able to process the message as intended.

I've looked at the code in `asyncio-mqtt` but it's unclear to me how to prevent this condition. Any advice is greatly appreciated.

For background, the bot and mosquito are running with docker-compose, and are on the same docker network. My connection and message subscription loop is identical to the code in [the test](https://github.com/sbtinstruments/asyncio-mqtt/blob/master/tests/reconnect.py#L17-L30).

**Comment by frederikaalund** [comment]
Thanks for submitting this issue. Let me have a look.

> There is very little activity, so typically there is hours or even days between messages received.

I've never tested the library to this extreme, so it's good to have some data on the matter. :+1:

I'm guessing that the connection is lost due to the long wait. This is curious, though, since I was under the impression that paho-mqtt continuously pings the server to keep the connection alive. This is exactly what the `keepalive` parameter is supposed to do.

I have some questions:
 * Do you see a ping message every 60 seconds in the mqtt broker log?
 * Does the issue occur with paho-mqtt itself (without asyncio-mqtt)?

Again, thanks for reporting this issue. :+1:

**Comment by SimplicityGuy** [comment]
I've been debugging this a bit, but still can't figure out why this is happening. There doesn't seem to be a rhyme or reason why the initial message doesn't get processed. I thought originally it was because of the length of time of the previous connection, however, looking at my logs if I add messages pretty close together, the first message never is sent. I see the same disconnect status.

I do not see a reconnect every 60s. My log shows the initial connection as:
```
mqtt       | 2020-12-25T23:12:06: New client connected from 172.26.0.2 as listener (p2, c1, k60).
```

The originating message is sent with `paho-mqtt` itself. I'm thinking of redoing that code using `asyncio-mqtt`. Originally just felt like overkill to send a message since I didn't need the asynchronous behavior.

**Comment by frederikaalund** [comment]
Thanks for the follow-up with additional information. :+1: 

> There doesn't seem to be a rhyme or reason why the initial message doesn't get processed. I thought originally it was because of the length of time of the previous connection, however, looking at my logs if I add messages pretty close together, the first message never is sent.

These kind of issues are especially difficult to debug. :) Right now, the culprit can be either of the following:
 1. An issue in asyncio-mqtt (this repo)
 2. An issue in paho-mqtt (our dependency)
 3. An issue in your code itself
 4. An issue with your local setup (MQTT broker, network, etc.)

I'd like to have a go at this issue but the problem space (1–4 above) is way too big. Can you reproduce the issue somehow? Preferably with an independent code example that triggers the bug reliably.

---

Anyhow, here are some things that you can try:
 * Increase the [QoS setting](https://en.wikipedia.org/wiki/MQTT#Quality_of_service_(QoS)) to exactly_once.
 * Use `logger.info("Message %s %s", message.topic, message.payload)` instead of `logger.info("Message %s %s", message.topic, message.payload.decode())`. The latter (from the test code) assumes UTF8-encoded payload. Maybe this is not the case in your scenario?
 * Thinking a bit further about the previous point: Maybe something inside your `async for message in messages` loop body raises an exception. You can try to simplify the code inside the `for` loop to simply print the message for the purpose of testing.

**Comment by SimplicityGuy** [comment]
Sorry it's taken so long for me to get back to this. I've swapped my `paho-mqtt` code out entirely for `asyncio-mqtt`. (Actually saved me a few lines of code, yay for less code and more readable code!)

However, the problem persists. Originally I filed it as a long time with an open connection, but instead I'd like to amend this to say "the first call to". I still don't quite know why this happens, since that code path is hit again once a job completes too, and the second time through it works just fine.

I suspect it could still be any of 1 through 3. I doubt it's (4) since I am able to send messages to the topic and receive/be notified of new messages. I'm going to spend a little more time debugging this.

**Comment by frederikaalund** [comment]
> Sorry it's taken so long for me to get back to this. 

That's completely fine. No hurry. At least not on my part. :)

> I've swapped my paho-mqtt code out entirely for asyncio-mqtt. (Actually saved me a few lines of code, yay for less code and more readable code!)

Glad to hear that. :+1: 

> However, the problem persists. Originally I filed it as a long time with an open connection, but instead I'd like to amend this to say "the first call to". I still don't quite know why this happens, since that code path is hit again once a job completes too, and the second time through it works just fine.

Ah, well, that changes things. It is indeed very strange that it behaves differently the first time only. Can you show me your code somehow? Post it here maybe?

> I suspect it could still be any of 1 through 3. I doubt it's (4) since I am able to send messages to the topic and receive/be notified of new messages. I'm going to spend a little more time debugging this.

Option 4 is a long shot. But let's not leave it out completely just yet.

**Comment by pichwo** [comment]
... looking amazed on this wonderful code ...

i am trying to write sort of topic-translator/state machine to connect crouton dashboard(s) to tinkerforge mqtt-sources...
one can imagine 

- 1 mqtt connection per dashboard which might connect to the broker and issue requests
and 
- 1 connection per data-source - which must be started and stopped as subprocess and will connect to the broker and serve requests

imy app n the middle keeping states and translating topics ;-)

so lots of parallel tasks which have to run endlessly and be kept happy ..
i think this is an ideal scenario for as yncio ...

**Comment by frederikaalund** [comment]
> hope you solve the issue ... - might it be some race condition or broken assumption about ordering of events ?
is there some documentation or at least some big picture of the overall design ?

Hopefully this becomes apparent when we get the repro code. :)

**Comment by SimplicityGuy** [comment]
[repro.zip](https://github.com/sbtinstruments/asyncio-mqtt/files/5818089/repro.zip)

Attached is a repro, however, using that repro I'm *unable* to repro the issue!

What I left out here due to Let's Encrypt certs, and this is a biggie, is the Traefik setup. My setup with Traefik and Let's Encrypt is identical to that of https://gist.github.com/natcl/ed8253a34e7b87d879baabeba82cb846. I tried to add in Traefik without Let's Encrypt, but didn't get it to work. Figured I'd point you to that setup in case you want to give this a try and add in Traefik along with some dummy sub-domains that you control.

So it's possible that there's a weird "first time only" issue when listening to messages in `bot.py` using Traefik. But then why it works the next time is still a mystery.

**Comment by SimplicityGuy** [comment]
I've dug into this a bit more. Reading the documentation on MQTT and Mosquitto, it seems that each connection client id should be unique. So I've appended a `uuid4().hex` to the client id. This way each connection made will have a unique id. With that fix, I've not been able to reproduce this issue.

**Comment by frederikaalund** [comment]
Glad that you found a solution to this. :tada: 

The fix has some ties to #48. If we encourage the user to reuse the `Client` instance, then they will also reuse the ID. I think that change solves both these issues.
